pkg/reduce/readloop/readloop.go

```diff
@@ -115,87 +115,156 @@ func (rl *ReadLoop) Startup(ctx context.Context) error {
-				// identify partition for message
-				partitionID := partition.ID{
-					Start: kw.StartTime(),
-					End:   kw.EndTime(),
-					Key:   m.Key,
-				}
+	q := rl.associatePBQAndPnF(ctx, p)
+
+	err := wait.ExponentialBackoff(pbqWriteBackoff, func() (done bool, err error) {
+		rErr := q.Write(context.Background(), m)
+		if rErr != nil {
+			rl.log.Errorw("Failed to write message", zap.Any("msgOffSet", m.ReadOffset.String()), zap.String("partitionID", p.String()), zap.Error(rErr))
+			// no point retrying if ctx.Done has been invoked
+			select {
+			case <-ctx.Done():
+				// no point in retrying after we have been asked to stop.
+				return false, ctx.Err()
+			default:
+				// keep retrying
+				return false, nil
+			}
 
-				q := rl.associatePBQAndPnF(ctx, partitionID)
-				// write the message to PBQ
-				attempt := 0
-				timeoutErr := wait.ExponentialBackoff(pbqWriteBackoff, func() (done bool, err error) {
-					// FIXME(p0): temporarily fixing happy path
-					rErr := q.Write(context.Background(), m)
-					attempt += 1
-					if rErr != nil {
-						rl.log.Errorw("Failed to write message", zap.Any("msgOffSet", m.ReadOffset.String()), zap.String("partitionID", partitionID.String()), zap.Any("attempt", attempt), zap.Error(rErr))
-						return false, nil
-					}
-					return true, nil
-				})
-
-				if timeoutErr != nil {
-					rl.log.Errorw("Not able to write the messages to all partitions. To avoid message loss, this message will not be acked", zap.String("msgOffSet", m.ReadOffset.String()))
-					return
-				}
+		}
+		// happy path
+		rl.log.Debugw("Successfully wrote the message", zap.String("msgOffSet", m.ReadOffset.String()), zap.String("partitionID", p.String()))
+		return true, nil
+	})
 
-			}
+	return err
+}
 
-			// Ack the message to ISB
-			attempt := 0
-			timeoutErr := wait.ExponentialBackoff(pbqWriteBackoff, func() (done bool, err error) {
-				rErr := m.ReadOffset.AckIt()
-				attempt += 1
-				if rErr != nil {
-					rl.log.Errorw("Failed to ack message", zap.String("msgOffSet", m.ReadOffset.String()), zap.Int("attempt", attempt), zap.Error(rErr))
+// ackMessages acks messages. Retries until it can succeed or ctx.Done() happens.
+func (rl *ReadLoop) ackMessages(ctx context.Context, messages []*isb.ReadMessage) error {
+	var err error
+	var ackBackoff = wait.Backoff{
+		Steps:    math.MaxInt,
+		Duration: 1 * time.Second,
+		Factor:   1.5,
+		Jitter:   0.1,
+	}
+	// Ack the message to ISB
+	for _, m := range messages {
+		err := wait.ExponentialBackoff(ackBackoff, func() (done bool, err error) {
+			rErr := m.ReadOffset.AckIt()
+			if rErr != nil {
+				rl.log.Errorw("Failed to ack message", zap.String("msgOffSet", m.ReadOffset.String()), zap.Error(rErr))
+				// no point retrying if ctx.Done has been invoked
+				select {
+				case <-ctx.Done():
+					// no point in retrying after we have been asked to stop.
+					return false, ctx.Err()
+				default:
+					// keep retrying
 					return false, nil
 				}
```

Comment:
This can be achieved by using ExponentialBackoffWithContext

Reply:
`ExponentialBackoffWithContext` will not work because it will return without executing anything if `ctx.Done()` has been closed. https://github.com/kubernetes/apimachinery/blob/v0.25.4/pkg/util/wait/wait.go#L734-L738

Reply:
yeah makes sense